# Explore enron data

In [31]:
import pickle

In [32]:
ls

choose_your_own/         feature_selection/  pca/            tools/
datasets_questions/      final_project/      README.md       validation/
decision_tree/           k_means/            regression/
enron-exploration.ipynb  naive_bayes/        svm/
evaluation/              outliers/           text_learning/


In [33]:
enron_data = pickle.load(open("final_project/final_project_dataset.pkl", "r"))

In [35]:
# load dataset into dataframe
import pandas as pd

df_clean = pd.DataFrame(enron_data)

In [36]:
df_clean.head()

ALLEN PHILLIP K BADUM JAMES P  \
bonus                              4175000           NaN   
deferral_payments                  2869717        178980   
deferred_income                   -3081055           NaN   
director_fees                          NaN           NaN   
email_address      phillip.allen@enron.com           NaN   

                           BANNANTINE JAMES M BAXTER JOHN C  \
bonus                                     NaN       1200000   
deferral_payments                         NaN       1295738   
deferred_income                         -5104      -1386055   
director_fees                             NaN           NaN   
email_address      james.bannantine@enron.com           NaN   

                        BAY FRANKLIN R BAZELIDES PHILIP J  \
bonus                           400000                NaN   
deferral_payments               260455             684694   
deferred_income                -201641                NaN   
director_fees                      NaN                NaN   
email_address      frank.bay@enron.com                NaN   

                           BECK SALLY W      BELDEN TIMOTHY N BELFER ROBERT  \
bonus                            700000               5249999           NaN   
deferral_payments                   NaN               2144013       -102500   
deferred_income                     NaN              -2334434           NaN   
director_fees                       NaN                   NaN          3285   
email_address      sally.beck@enron.com  tim.belden@enron.com           NaN   

                             BERBERIAN DAVID    ...     \
bonus                                    NaN    ...      
deferral_payments                        NaN    ...      
deferred_income                          NaN    ...      
director_fees                            NaN    ...      
email_address      david.berberian@enron.com    ...      

                             WASAFF GEORGE       WESTFAHL RICHARD K  \
bonus                               325000                      NaN   
deferral_payments                   831299                      NaN   
deferred_income                    -583325                   -10800   
director_fees                          NaN                      NaN   
email_address      george.wasaff@enron.com  dick.westfahl@enron.com   

                  WHALEY DAVID A      WHALLEY LAWRENCE G  \
bonus                        NaN                 3000000   
deferral_payments            NaN                     NaN   
deferred_income              NaN                     NaN   
director_fees                NaN                     NaN   
email_address                NaN  greg.whalley@enron.com   

                        WHITE JR THOMAS E WINOKUR JR. HERBERT S  \
bonus                              450000                   NaN   
deferral_payments                     NaN                   NaN   
deferred_income                       NaN                -25000   
director_fees                         NaN                108579   
email_address      thomas.white@enron.com                   NaN   

                             WODRASKA JOHN WROBEL BRUCE  \
bonus                                  NaN          NaN   
deferral_payments                      NaN          NaN   
deferred_income                        NaN          NaN   
director_fees                          NaN          NaN   
email_address      john.wodraska@enron.com          NaN   

                           YEAGER F SCOTT YEAP SOON  
bonus                                 NaN       NaN  
deferral_payments                     NaN       NaN  
deferred_income                       NaN       NaN  
director_fees                         NaN       NaN  
email_address      scott.yeager@enron.com       NaN  

[5 rows x 146 columns]

In [37]:
df_clean.shape

(21, 146)

so we have 21 features and 146 persons of interests, but the order is weird, let's transpose it to have the features of each person in a row

In [43]:
df = df_clean.transpose()

In [44]:
df.head()

bonus deferral_payments deferred_income director_fees  \
ALLEN PHILLIP K     4175000           2869717        -3081055           NaN   
BADUM JAMES P           NaN            178980             NaN           NaN   
BANNANTINE JAMES M      NaN               NaN           -5104           NaN   
BAXTER JOHN C       1200000           1295738        -1386055           NaN   
BAY FRANKLIN R       400000            260455         -201641           NaN   

                                 email_address exercised_stock_options  \
ALLEN PHILLIP K        phillip.allen@enron.com                 1729541   
BADUM JAMES P                              NaN                  257817   
BANNANTINE JAMES M  james.bannantine@enron.com                 4046157   
BAXTER JOHN C                              NaN                 6680544   
BAY FRANKLIN R             frank.bay@enron.com                     NaN   

                   expenses from_messages from_poi_to_this_person  \
ALLEN PHILLIP K       13868          2195                      47   
BADUM JAMES P          3486           NaN                     NaN   
BANNANTINE JAMES M    56301            29                      39   
BAXTER JOHN C         11200           NaN                     NaN   
BAY FRANKLIN R       129142           NaN                     NaN   

                   from_this_person_to_poi        ...         \
ALLEN PHILLIP K                         65        ...          
BADUM JAMES P                          NaN        ...          
BANNANTINE JAMES M                       0        ...          
BAXTER JOHN C                          NaN        ...          
BAY FRANKLIN R                         NaN        ...          

                   long_term_incentive    other    poi restricted_stock  \
ALLEN PHILLIP K                 304805      152  False           126027   
BADUM JAMES P                      NaN      NaN  False              NaN   
BANNANTINE JAMES M                 NaN   864523  False          1757552   
BAXTER JOHN C                  1586055  2660303  False          3942714   
BAY FRANKLIN R                     NaN       69  False           145796   

                   restricted_stock_deferred  salary shared_receipt_with_poi  \
ALLEN PHILLIP K                      -126027  201955                    1407   
BADUM JAMES P                            NaN     NaN                     NaN   
BANNANTINE JAMES M                   -560222     477                     465   
BAXTER JOHN C                            NaN  267102                     NaN   
BAY FRANKLIN R                        -82782  239671                     NaN   

                   to_messages total_payments total_stock_value  
ALLEN PHILLIP K           2902        4484442           1729541  
BADUM JAMES P              NaN         182466            257817  
BANNANTINE JAMES M         566         916197           5243487  
BAXTER JOHN C              NaN        5634343          10623258  
BAY FRANKLIN R             NaN         827696             63014  

[5 rows x 21 columns]

let's count the number of persons of interest

In [47]:
pois = df[df.poi == True]

In [48]:
# number of persons of interest
len(pois)

18

let's count the total number of pois

In [49]:
total_num_pois = 0
with open('final_project/poi_names.txt', 'r') as f:
    for line in f:
        if any(keyword in line for keyword in ["(y)","(n)"]):
            total_num_pois += 1

In [50]:
total_num_pois

35

in total there are 35 persons of interest

What is the total value of the stock belonging to James Prentice?

In [54]:
df_clean.columns.values

array(['ALLEN PHILLIP K', 'BADUM JAMES P', 'BANNANTINE JAMES M',
       'BAXTER JOHN C', 'BAY FRANKLIN R', 'BAZELIDES PHILIP J',
       'BECK SALLY W', 'BELDEN TIMOTHY N', 'BELFER ROBERT',
       'BERBERIAN DAVID', 'BERGSIEKER RICHARD P', 'BHATNAGAR SANJAY',
       'BIBI PHILIPPE A', 'BLACHMAN JEREMY M', 'BLAKE JR. NORMAN P',
       'BOWEN JR RAYMOND M', 'BROWN MICHAEL', 'BUCHANAN HAROLD G',
       'BUTTS ROBERT H', 'BUY RICHARD B', 'CALGER CHRISTOPHER F',
       'CARTER REBECCA C', 'CAUSEY RICHARD A', 'CHAN RONNIE',
       'CHRISTODOULOU DIOMEDES', 'CLINE KENNETH W', 'COLWELL WESLEY',
       'CORDES WILLIAM R', 'COX DAVID', 'CUMBERLAND MICHAEL S',
       'DEFFNER JOSEPH M', 'DELAINEY DAVID W', 'DERRICK JR. JAMES V',
       'DETMERING TIMOTHY J', 'DIETRICH JANET R', 'DIMICHELE RICHARD G',
       'DODSON KEITH', 'DONAHUE JR JEFFREY M', 'DUNCAN JOHN H',
       'DURAN WILLIAM D', 'ECHOLS JOHN B', 'ELLIOTT STEVEN',
       'FALLON JAMES B', 'FASTOW ANDREW S', 'FITZGERALD JAY L',
       'FOW

ok, obviously we need to use family and firstname written in upper case

In [56]:
df.loc["PRENTICE JAMES"]["total_stock_value"]

1095040

The total value of stock of James Prentice is 1095040.

What’s the value of stock options exercised by Jeffrey K Skilling?

In [58]:
df.loc["SKILLING JEFFREY K"]["exercised_stock_options"]

19250000

The value of stock options exercised by Jeffery K Skilling is: 19250000.

todo
quizzes: 21

Of these three individuals (Lay, Skilling and Fastow), who took home the most money (largest value of “total_payments” feature)?

How much money did that person get?

In [87]:
df.loc[["SKLLING JEFFERY K", "LAY KENNETH L", "FASTOW ANDREW S"]]["total_payments"]

SKLLING JEFFERY K          NaN
LAY KENNETH L        103559793
FASTOW ANDREW S        2424083
Name: total_payments, dtype: object

Kenneth Lay was the guy with an amount of 103559793.

How many folks in this dataset have a quantified salary? What about a known email address?
Start Quiz

In [99]:
df.columns.values

array(['bonus', 'deferral_payments', 'deferred_income', 'director_fees',
       'email_address', 'exercised_stock_options', 'expenses',
       'from_messages', 'from_poi_to_this_person',
       'from_this_person_to_poi', 'loan_advances', 'long_term_incentive',
       'other', 'poi', 'restricted_stock', 'restricted_stock_deferred',
       'salary', 'shared_receipt_with_poi', 'to_messages',
       'total_payments', 'total_stock_value'], dtype=object)

In [112]:
# does not work here, maybe isnull function had a bug in that version?
#len(df[~df.salary.isnull()])

# using that alternativ, did not work neither
#len(df.salary.dropna())

# manual checking works, maybe 'NaN' here is of different type
len(df[df.salary != 'NaN'])

95

95 persons had a quantified salary.

In [119]:
len(df[df.email_address != 'NaN'])

111

In [123]:
df_0 = df.replace('NaN', 0)